politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+	// Prepare blob
+	be, err := convertBlobEntryFromAuthDetails(ad)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(treeID, *be)
+}
+
+func (p *ticketVotePlugin) auths(treeID int64) ([]ticketvote.AuthDetails, error) {
+	// Retrieve blobs
+	blobs, err := p.tstore.BlobsByDataDesc(treeID,
+		[]string{dataDescriptorAuthDetails})
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode blobs
+	auths := make([]ticketvote.AuthDetails, 0, len(blobs))
+	for _, v := range blobs {
+		a, err := convertAuthDetailsFromBlobEntry(v)
+		if err != nil {
+			return nil, err
+		}
+		auths = append(auths, *a)
+	}
+
+	// Sanity check. They should already be sorted from oldest to
+	// newest.
+	sort.SliceStable(auths, func(i, j int) bool {
+		return auths[i].Timestamp < auths[j].Timestamp
+	})
+
+	return auths, nil
+}
+
+func (p *ticketVotePlugin) voteDetailsSave(treeID int64, vd ticketvote.VoteDetails) error {
+	// Prepare blob
+	be, err := convertBlobEntryFromVoteDetails(vd)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(treeID, *be)
+}
+
+func (p *ticketVotePlugin) voteDetails(treeID int64) (*ticketvote.VoteDetails, error) {
+	// Retrieve blobs
+	blobs, err := p.tstore.BlobsByDataDesc(treeID,
+		[]string{dataDescriptorVoteDetails})
+	if err != nil {
+		return nil, err
+	}
+	switch len(blobs) {
+	case 0:
+		// A vote details does not exist
+		return nil, nil
+	case 1:
+		// A vote details exists; continue
+	default:
+		// This should not happen. There should only ever be a max of
+		// one vote details.
+		return nil, fmt.Errorf("multiple vote details found (%v) on %x",
+			len(blobs), treeID)
+	}
+
+	// Decode blob
+	vd, err := convertVoteDetailsFromBlobEntry(blobs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return vd, nil
+}
+
+func (p *ticketVotePlugin) voteDetailsByToken(token []byte) (*ticketvote.VoteDetails, error) {
+	reply, err := p.backend.PluginRead(token, ticketvote.PluginID,
+		ticketvote.CmdDetails, "")
+	if err != nil {
+		return nil, err
+	}
+	var dr ticketvote.DetailsReply
+	err = json.Unmarshal([]byte(reply), &dr)
+	if err != nil {
+		return nil, err
+	}
+	return dr.Vote, nil
+}
+
+func (p *ticketVotePlugin) castVoteSave(treeID int64, cv ticketvote.CastVoteDetails) error {
+	// Prepare blob
+	be, err := convertBlobEntryFromCastVoteDetails(cv)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(treeID, *be)
+}
+
+func (p *ticketVotePlugin) voteResults(treeID int64) ([]ticketvote.CastVoteDetails, error) {
+	// Retrieve blobs
+	desc := []string{
+		dataDescriptorCastVoteDetails,
+		dataDescriptorVoteCollider,
+	}
+	blobs, err := p.tstore.BlobsByDataDesc(treeID, desc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode blobs. A cast vote is considered valid only if the vote
+	// collider exists for it. If there are multiple votes using the
+	// same ticket, the valid vote is the one that immediately precedes
+	// the vote collider blob entry.
+	var (
```

Comment:
Is this a guarantee or is it "precedesish"?

Reply:
From a tlog standpoint its "precedesish" but from a plugin standpoint it is precedes.  Plugins can't fetch arbitrary leaves from tlog.  They have to request specific leaf types, so when they request the vote leaves the valid vote will always precede the vote collider.  The plugin will not know about any leaves that make it in between the cast vote and the vote collider.